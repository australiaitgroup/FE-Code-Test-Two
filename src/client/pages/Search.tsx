import * as React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import { doSearch, handleInputChangeAction } from '../actions/search';
import Layout from '../components/Layout';
import Pagination from '../components/Pagination';
import SearchMetadata from '../components/search/SearchMetadata';
import SearchBar from '../components/search/SearchBar';
import SearchResult from '../components/search/SearchResult';
import { IState } from '../../types/stateAndAction';
import {
    ISearchDispatchProps,
    ISearchProps,
    ISearchStateProps,
} from './interfaces/search.interface';

const mapStateToProps = (state: IState): ISearchStateProps => ({
    error: state.search.error,
    hits: state.search.result.hits,
    isFetching: state.search.isFetching,
    nbHits: state.search.result.nbHits,
    nbPages: state.search.result.nbPages,
    page: state.search.page,
    settings: state.settings,
});

const mapDispatchToProps = (dispatch): ISearchDispatchProps => ({
    changePage: page => dispatch(doSearch(page)),
    handleInputChange: input => dispatch(handleInputChangeAction(input)),
    search: () => dispatch(doSearch()),
});

class Search extends React.Component<ISearchProps> {
    componentDidMount() {
        this.props.search();
    }

    render() {
        const {
            changePage,
            error,
            handleInputChange,
            hits,
            isFetching,
            nbHits,
            nbPages,
            page,
            search,
            settings,
        } = this.props;

        return (
            <Layout>
                <SearchBar
                    handleInputChange={handleInputChange}
                    search={search}
                />
                <SearchMetadata nbHits={nbHits} settings={settings} />
                <SearchResult
                    error={error}
                    hits={hits}
                    isFetching={isFetching}
                    nbHits={nbHits}
                />
                <Pagination
                    handlePageChange={changePage}
                    nbPages={nbPages}
                    page={page}
                />
            </Layout>
        );
    }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Search));
