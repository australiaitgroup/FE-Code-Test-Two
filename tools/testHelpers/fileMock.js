// Used by Jest: refer to https://jestjs.io/docs/en/webpack.html

const path = require('path');

module.exports = {
    process(src, filename) {
        return `module.exports = ${JSON.stringify(path.basename(filename))};`;
    },
};
