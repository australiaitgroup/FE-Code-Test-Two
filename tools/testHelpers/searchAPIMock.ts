import * as nock from 'nock';

export default nock('https://hn.algolia.com/api/v1');
