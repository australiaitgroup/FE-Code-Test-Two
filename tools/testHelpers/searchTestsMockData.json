{
  "searchResult": {
    "hits": [
      {
        "created_at": "2015-06-03T19:45:33.000Z",
        "title": null,
        "url": null,
        "author": "pingec",
        "points": null,
        "story_text": null,
        "comment_text": "test",
        "num_comments": null,
        "story_id": 9654758,
        "story_title": "How Well Can You Hear Audio Quality?",
        "story_url": "http://www.npr.org/sections",
        "parent_id": 9654758,
        "created_at_i": 1433360733,
        "objectID": 32223433,
        "_highlightResult": {
          "author": {
            "matchLevel": "none",
            "matchedWords": [],
            "value": "pingec"
          },
          "comment_text": {
            "fullyHighlighted": false,
            "matchLevel": "full",
            "matchedWords": ["testttd"],
            "value": "Similar test: <a href='http://test.tidalhifi.com'"
          },
          "story_title": {
            "matchLevel": "none",
            "matchedWords": [],
            "value": "How Well Can You Hear Audio Quality?"
          },
          "story_url": {
            "matchLevel": "none",
            "matchedWords": [],
            "value": "http://www.npr.org/sections/therecord/2015/06/02/"
          }
        },
        "_tags": ["comment", "author_pingec", "story_9654758"]
      }
    ],
    "nbHits": 1,
    "page": 0,
    "nbPages": 1,
    "hitsPerPage": 20,
    "processingTimeMS": 10,
    "exhaustiveNbHits": true,
    "query": "test",
    "params": "advancedSyntax=true&analytics=false&hitsPerPage=20&query=testttd&tags=%28story%2Ccomment%29"
  }
}
