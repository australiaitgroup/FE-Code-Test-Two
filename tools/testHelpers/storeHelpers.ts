import { Store } from 'redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import reducers from '../../src/client/reducers';
import { IState } from '../../src/types/stateAndAction';

export function getInitialState(): IState {
    return reducers(undefined, { type: null });
}

export function createMockStore(initialState: IState): Store {
    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);
    return mockStore(initialState);
}

export function deepClone(toBeCloned: any): any {
    return JSON.parse(JSON.stringify(toBeCloned));
}
