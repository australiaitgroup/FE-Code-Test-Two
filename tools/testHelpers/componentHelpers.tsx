import * as React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import * as renderer from 'react-test-renderer';

import { createMockStore } from './storeHelpers';

export function renderComponent(children, storeData) {
    const store = createMockStore(storeData);

    return renderer.create(
        <Provider store={store}>
            <BrowserRouter>
                {children}
            </BrowserRouter>
        </Provider>);
}
