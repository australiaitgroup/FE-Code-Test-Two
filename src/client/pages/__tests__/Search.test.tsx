import * as React from 'react';

import Search from '../Search';
import { renderComponent } from '../../../../tools/testHelpers/componentHelpers';
import apiMock from '../../../../tools/testHelpers/searchAPIMock';
import { getInitialState } from '../../../../tools/testHelpers/storeHelpers';

it('Search renders correctly', async () => {
    apiMock.get('/search?query=&tags=(story,comment)&hitsPerPage=20')
        .reply(200, {});

    const component = await renderComponent(<Search />, getInitialState());
    const result = component.toJSON();
    expect(result).toMatchSnapshot();
});
