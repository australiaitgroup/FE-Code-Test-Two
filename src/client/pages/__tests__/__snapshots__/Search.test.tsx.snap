// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<div
  className="ui segment pushable fullHeight flex inheritBg"
>
  <div
    className="ui inverted vertical labeled icon ui overlay right sidebar menu"
  >
    <div
      className="item"
      onClick={[Function]}
    >
      <p>
        Search type
      </p>
      <div
        aria-expanded={false}
        className="ui selection dropdown"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        role="listbox"
        tabIndex={0}
      >
        <div
          aria-live="polite"
          className="text"
          role="alert"
        >
          All
        </div>
        <i
          aria-hidden="true"
          className="dropdown icon"
          onClick={[Function]}
        />
        <div
          className="menu transition"
        >
          <div
            aria-checked={true}
            aria-selected={true}
            className="active selected item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              All
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Stories
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Comments
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
    >
      <p>
        Sort order
      </p>
      <div
        aria-expanded={false}
        className="ui selection dropdown"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        role="listbox"
        tabIndex={0}
      >
        <div
          aria-live="polite"
          className="text"
          role="alert"
        >
          Popularity
        </div>
        <i
          aria-hidden="true"
          className="dropdown icon"
          onClick={[Function]}
        />
        <div
          className="menu transition"
        >
          <div
            aria-checked={true}
            aria-selected={true}
            className="active selected item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Popularity
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Date
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="horizontally fitted item"
      onClick={[Function]}
    >
      <p>
        Date range
      </p>
      <div
        aria-expanded={false}
        className="ui selection dropdown"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        role="listbox"
        tabIndex={0}
      >
        <div
          aria-live="polite"
          className="text"
          role="alert"
        >
          All time
        </div>
        <i
          aria-hidden="true"
          className="dropdown icon"
          onClick={[Function]}
        />
        <div
          className="menu transition"
        >
          <div
            aria-checked={true}
            aria-selected={true}
            className="active selected item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              All time
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Last 24h
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Past Week
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Past Month
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Past Year
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              Custom range
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
    >
      <p>
        Hits per page
      </p>
      <div
        aria-expanded={false}
        className="ui selection dropdown"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        role="listbox"
        tabIndex={0}
      >
        <div
          aria-live="polite"
          className="text"
          role="alert"
        >
          20
        </div>
        <i
          aria-hidden="true"
          className="dropdown icon"
          onClick={[Function]}
        />
        <div
          className="menu transition"
        >
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              10
            </span>
          </div>
          <div
            aria-checked={true}
            aria-selected={true}
            className="active selected item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              20
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              30
            </span>
          </div>
          <div
            aria-checked={false}
            aria-selected={false}
            className="item"
            onClick={[Function]}
            role="option"
            style={
              Object {
                "pointerEvents": "all",
              }
            }
          >
            <span
              className="text"
            >
              50
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      className="item"
      onClick={[Function]}
    >
      <button
        className="ui green mini button"
        onClick={[Function]}
        role="button"
      >
        Apply
      </button>
      <button
        className="ui mini button"
        onClick={[Function]}
        role="button"
      >
        Cancel
      </button>
    </div>
  </div>
  <div
    className="ui teal huge borderless inverted noMarginTop menu"
  >
    <div
      className="menu"
      widths={2}
    >
      <div
        className="header item"
        onClick={[Function]}
      >
        <img
          className="ui mini right spaced image"
        />
        Hacker news search
      </div>
    </div>
    <div
      className="right menu"
    >
      <button
        className="ui teal button"
        onClick={[Function]}
        role="button"
      >
        <i
          aria-hidden="true"
          className="setting large icon"
        />
      </button>
    </div>
  </div>
  <div
    className="pusher flex flexGrow"
    onClick={[Function]}
  >
    <div
      className="ui container"
    >
      <div
        className="ui fluid action left icon input"
      >
        <input
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="Search..."
          type="text"
        />
        <i
          aria-hidden="true"
          className="search icon"
        />
        <button
          className="ui teal button"
          onClick={[Function]}
          role="button"
        >
          Search
        </button>
      </div>
      <div
        className="ui segment"
      >
        <div
          className="ui divided items"
        />
      </div>
    </div>
    <div
      className="ui vertical segment fixDown"
    >
      <div
        className="ui center aligned container"
      >
        <div>
          <div
            className="item"
            onClick={[Function]}
          >
            Copyright © 
            2018
             Joey Zheng
          </div>
          <div
            className="item"
            onClick={[Function]}
          >
            v 1.0
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
